Task 3/Task 3.2/Task3.2.1.Library/DynamicArray.cs

```diff
@@ -0,0 +1,281 @@
+        /// </summary>
+        public void Add(T value) // 4
+        {
+            IncreaseCapacity(Length + 1);
+
+            _array[Length] = value;
+            Length+= 1;
+        }
+
+        /// <summary>
+        /// Метод AddRange, добавляющий в конец массива содержимое коллекции, реализующей
+        /// интерфейс IEnumerable<T>.Обратите внимание, метод должен корректно учитывать число
+        /// элементов в коллекции с тем, чтобы при необходимости расширения массива делать это
+        /// только один раз вне зависимости от числа элементов в добавляемой коллекции.
+        /// </summary>
+        public void AddRange(IEnumerable<T> sourse) // 5
+        {
+            IncreaseCapacity(Length + sourse.Count());
+
+            foreach(var item in sourse)
+            {
+                _array[Length++] = item;
+            }
+        } 
+
+
+        /// <summary>
+        /// Метод Remove, удаляющий из коллекции указанный элемент. Метод должен возвращать
+        /// true, если удаление прошло успешно и false в противном случае.При удалении элементов
+        /// реальная ёмкость массива не должна уменьшаться.
+        /// </summary>
+        public bool Remove(T value) // 6
+        {
+            for (int i = 0; i < Length; i++)
+            {
+                if (_array[i].Equals(value))
+                {
+                    LeftShift(i + 1, 1);
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Метод Insert, позволяющий добавить элемент в произвольную позицию массива (обратите
+        /// внимание, может потребоваться расширить массив). Метод должен возвращать true, если
+        /// добавление прошло успешно и false в противном случае.При выходе за границу массива
+        /// должно генерироваться исключение ArgumentOutOfRangeException.
+        /// </summary>
+        public bool Insert(int index, T item) // 7
+        {
+            if (index < 0 || index > Length)
+                throw new ArgumentOutOfRangeException();
+
+            RightShift(index, 1);
+            _array[index] = item;
+            return true;
+        }
+
+
+        /// <summary>
+        /// Добавить метод ToArray, возвращающий новый массив (обычный), содержащий все
+        /// содержащиеся в текущем динамическом массиве объекты.
+        /// </summary>
+        public T[] ToArray() // 4*
+        {
+            T[] newArray = new T[Length];
+
+            Array.Copy(_array, newArray, Length);
+
+            return newArray;
+        }
+
+
+        /// <summary>
+        /// Метод, реализующий интерфейс IEnumerable.
+        /// </summary>
+        IEnumerator IEnumerable.GetEnumerator() => (IEnumerator)GetEnumerator(); // 10
+
+        /// <summary>
+        /// Метод, реализующий интерфейс IEnumerable<T>.
+        /// </summary>
+        public IEnumerator<T> GetEnumerator() => new DynamicArrayEnum<T>(this); // 10
```

Comment:
You could write a method with `yield return` here - such methods are turned into iterators (enumerators) on compile